Fix databaseSupportsUTF8MB4

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Fix databaseSupportsUTF8MB4

Before
----------------------------------------
``databaseSupportsUTF8MB4`` always returns false

After
----------------------------------------
Only returns false when false

Technical Details
----------------------------------------
This function has been wrong since it was added

Comments
----------------------------------------
stripos returns 0 on success, not true.

eileenmcnaughton committed:
Fix databaseSupportsUTF8MB4

stripos returns 0 on success, not true.

seamuslee001 commented on CRM/Core/BAO/SchemaHandler.php:
```diff
@@ -920,7 +920,7 @@ public static function getInUseCollation(): string {
   public static function databaseSupportsUTF8MB4(): bool {
     if (!isset(\Civi::$statics[__CLASS__][__FUNCTION__])) {
-      \Civi::$statics[__CLASS__][__FUNCTION__] = stripos(self::getInUseCollation(), 'utf8mb4') === TRUE;
+      \Civi::$statics[__CLASS__][__FUNCTION__] = stripos(self::getInUseCollation(), 'utf8mb4') === 0;
```
@eileenmcnaughton what if instead it was `!== FALSE`

eileenmcnaughton replied:
Then the function would be wrong. The exact value returned when the string starts with 'utf8mb4' is 0 - because that is the position it holds in the string

seamuslee001 commented:
hmm I think the test fail relates

```
api_v3_ContactTest::testEmojiInWhereClause
Exception: PEAR_Exception: "DB Error: unknown error"
 * ERROR TYPE: DB_Error
 * ERROR CODE: -1
 * ERROR MESSAGE: DB Error: unknown error
 * ERROR MODE: 16
 * ERROR USERINFO: 
        ALTER TABLE civicrm_contact MODIFY COLUMN
        `first_name` VARCHAR(64) CHARACTER SET utf8 COLLATE utf8_unicode_ci DEFAULT NULL COMMENT 'First Name.',
        CHARSET utf8
       [nativecode=1283 ** Column 'last_name' cannot be part of FULLTEXT index]
```

eileenmcnaughton commented:
odd huh! I guess the test wasn't actually running because this was wrong & short-circuiting

eileenmcnaughton commented:
@seamuslee001 I dug into that test - it wasn't working before in that the database was never being updated to NOT support utf8mb4 because this function was wrong. Consequently it was giving a false pass. In fact this api call does fail fatally under apiv3 (but not v4).

I don't think it was ever fixed for v3 - the bugs just fooled us into thinking it was. However, I think worrying too much about  increasing support for utf8 is a ship that has sailed so I fixed the test to be apiv4 only

seamuslee001 commented:
@eileenmcnaughton same test failure again, I think the problem is that the full text mode hasn't been disabled I think

eileenmcnaughton commented:
@seamuslee001 oh yeah - i didn't get that error locally - I got a different one - but I ran the test in isolation so I guess I need to figure out where it gets enabled

eileenmcnaughton commented:
There seem to be multiple flaws in the cleanup of the INNODB full text indices. I think the query in https://github.com/civicrm/civicrm-core/pull/20912 will find them better than the existing one. However, I think by switching to using middle_name rather than first_name in this test we can bypass being blocked on that cleanup

eileenmcnaughton commented:
@seamuslee001 I've ripped out the part of the test that gets it to partially mimic a non-utf8mb4 supporting db on a utf8mb4 db - it seems that bit was always flawed & it won't play nice with full text indexes having a mismatch of field types in the table. Despite my best efforts I can't seem to fully purge those indexes & when I DO get this to run it messes up the db in some other way.

All of which is a way of saying I think we should rely on the matrix having a range of supported mysql configurations & not try to imitate one in this test. The test was always intended to age out

eileenmcnaughton commented on tests/phpunit/api/v3/ContactTest.php:
```diff
@@ -5194,29 +5194,13 @@ protected function doMerge(bool $isReverse = FALSE) {
-   * @throws \CRM_Core_Exception
    */
   public function testEmojiInWhereClause(): void {
-    $schemaNeedsAlter = CRM_Core_BAO_SchemaHandler::databaseSupportsUTF8MB4();
```
Was always FALSE before - so this test never worked except 'on the matrix'

seamuslee001 commented:
@eileenmcnaughton I did this locally on my local php8 and local php7.3 boxes and both seemed to pass with this by running `./tools/scripts/phpunit --filter 'testEmojiInWhereClause' 'api_v3_ContactTest'` 

```
diff --git a/tests/phpunit/api/v3/ContactTest.php b/tests/phpunit/api/v3/ContactTest.php
index 79b3e54caf..ac1b5c99dd 100644
--- a/tests/phpunit/api/v3/ContactTest.php
+++ b/tests/phpunit/api/v3/ContactTest.php
@@ -5194,12 +5194,14 @@ class api_v3_ContactTest extends CiviUnitTestCase {
    *
    * By default our DBs are not 🦉 compliant. This test will age
    * out when we are.
+   *
+   * @dataProvider versionThreeAndFour
    */
-  public function testEmojiInWhereClause(): void {
+  public function testEmojiInWhereClause(int $version): void {
+    $this->_apiversion = $version;
     $this->callAPISuccess('Contact', 'get', [
       'debug' => 1,
       'first_name' => '🦉Claire',
-      'version' => 4,
     ]);
   }
```

eileenmcnaughton commented:
@seamuslee001 yeah - the issue is the lines I removed attempted to 'pretend' that the db was in utf8 not utf4mb8 (if it was in utf8mb4) by changing one field over - but it never did as ``databaseSupportsUTF8MB4`` was always returning false - when I fixed that it failed because having a mix of utf8 & utf8mb4 fields in the table caused issues

seamuslee001 commented:
@eileenmcnaughton my point was more that I think this test can work in apiv3 and v4

eileenmcnaughton commented:
@seamuslee001 hmm - that would be odd since the code v3 goes through should fail on a mysql db that is in utf8

- you specified the php versions you tested on - but not the mysql charset - is your php 7.3 database on an old version of mysql with the old charset?

demeritcowboy commented:
For the purposes of testing the query logic if not the actual database, could you set `Civi::$statics['CRM_Core_BAO_SchemaHandler']['databaseSupportsUTF8MB4']` inside the test to test both true and false?

eileenmcnaughton commented:
@demeritcowboy so the issue is that we are testing that IF the database is not in utf8mb4 will it fatal crash if we try to search for an emoji in v4 api. We can only really test that if the field is utf8

demeritcowboy commented:
To test fully yes, but since you were suggesting scrapping it anyway then it seemed like the logic can still be tested.

I mean, if the database is utf8mb4, which it is on the test sites, then with the static set to true, it should return a contact matching the owl. With the static set to false, it should return blank, which I agree doesn't test utf8 directly, but tests the logic in the query of returning blank when it thinks it's utf8.

eileenmcnaughton commented:
@demeritcowboy technically I'm not talking about scrapping it - because it wasn't working anyway & was already bypassed. 

Our current cover is that the matrix is (should be) providing the range of versions / mysql set ups we cover & somewhere in there it should have a db version that is utf8 while we still support that - which seems like OK cover

demeritcowboy commented:
Sounds good.

So on this PR then it's just waiting to figure out how v3 could possibly have passed?

eileenmcnaughton commented:
@demeritcowboy yeah so in the context of CI we know the answer - ie it was never being tested. The question for @seamuslee001 is whether we still have configs in our matrix that are utf8 dbs

demeritcowboy commented:
Merging because the bug itself is clearly a bug, and can think about the test matrix cover for utf8 separately.
